src/middleware/devtools.ts

```diff
@@ -61,107 +74,190 @@ export const devtools =
   (
     set: CustomSetState,
     get: CustomGetState,
-    api: CustomStoreApi & StoreApiWithDevtools<S> & { dispatch?: unknown }
+    api: CustomStoreApi &
+      StoreApiWithDevtools<S> & {
+        dispatch?: unknown
+        dispatchFromDevtools?: boolean
+      }
   ): S => {
-    let extension
-    try {
-      extension =
-        (window as any).__REDUX_DEVTOOLS_EXTENSION__ ||
-        (window as any).top.__REDUX_DEVTOOLS_EXTENSION__
-    } catch {}
-
-    if (!extension) {
+    let devtoolsOptions =
+      options === undefined
+        ? { name: undefined }
+        : typeof options === 'string'
+        ? { name: options }
+        : options
+
+    let extensionConnector =
+      (window as any).__REDUX_DEVTOOLS_EXTENSION__ ??
+      (window as any).top.__REDUX_DEVTOOLS_EXTENSION__
+
+    if (!extensionConnector) {
       if (
         process.env.NODE_ENV === 'development' &&
         typeof window !== 'undefined'
       ) {
-        console.warn('Please install/enable Redux devtools extension')
+        console.warn(
+          '[zustand devtools middleware] Please install/enable Redux devtools extension'
+        )
       }
-      delete api.devtools
+
       return fn(set, get, api)
     }
-    const namedSet: NamedSet<S> = (state, replace, name) => {
+
+    let extension = Object.create(extensionConnector.connect(devtoolsOptions))
+    // We're using `Object.defineProperty` to set `prefix`, so if extensionConnector.connect
+    // returns the same reference we'd get cannot redefine property prefix error
+    // hence we `Object.create` to make a new reference
```

Comment:
Is this correct that `Object.defineProperty` and `Object.create` are only used for migration (warning messages) and we plan to remove them in v4?

Reply:
Yep, we'll remove them in v4